go-controller/pkg/ovnwebhook/nodeadmission.go

```diff
@@ -0,0 +1,169 @@
+	}
+	return &NodeAdmission{
+		annotationChecks: checks,
+		annotationKeys:   sets.New[string](maps.Keys(checks)...),
+	}
+}
+
+var _ admission.CustomValidator = &NodeAdmission{}
+
+func (p NodeAdmission) ValidateCreate(_ context.Context, _ runtime.Object) (warnings admission.Warnings, err error) {
+	// Ignore creation, the webhook is configured to only handle nodes/status updates
+	return nil, nil
+}
+
+func (p NodeAdmission) ValidateDelete(_ context.Context, _ runtime.Object) (warnings admission.Warnings, err error) {
+	// Ignore deletion, the webhook is configured to only handle nodes/status updates
+	return nil, nil
+}
+
+func (p NodeAdmission) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) (warnings admission.Warnings, err error) {
+	oldNode := oldObj.(*corev1.Node)
+	newNode := newObj.(*corev1.Node)
+
+	changes := mapDiff(oldNode.Annotations, newNode.Annotations)
+	changedKeys := maps.Keys(changes)
+
+	req, err := admission.RequestFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+	nodeName, isOVNKubeNode := ovnkubeNodeIdentity(req.UserInfo)
+
+	if !isOVNKubeNode {
+		// the user is not an ovnkube-node
+		return nil, nil
+	}
+
+	if newNode.Name != nodeName {
+		return nil, fmt.Errorf("ovnkube-node on node: %q is not allowed to modify nodes %q annotations", nodeName, newNode.Name)
+	}
+
+	// ovnkube-node is not allowed to change annotations outside of it's scope
+	if !p.annotationKeys.HasAll(changedKeys...) {
+		return nil, fmt.Errorf("ovnkube-node on node: %q is not allowed to set at least one of the annotations: %v", nodeName, changedKeys)
+	}
+
+	for _, key := range changedKeys {
+		if check := p.annotationChecks[key]; check != nil {
+			if err := check(changes[key], nodeName); err != nil {
+				return nil, fmt.Errorf("ovnkube-node on node: %q is not allowed to set %s on node %q: %v", nodeName, key, newNode.Name, err)
+			}
+		}
+	}
+
+	// Verify that nothing but the annotations changed.
+	// Since ovnkube-node only has the node/status permissions, it is enough to check .Status and .ObjectMeta only.
+	// Ignore .ManagedFields fields which are modified on every update.
+	oldNodeShallowCopy := oldNode
+	newNodeShallowCopy := newNode
+	oldNodeShallowCopy.Annotations = nil
+	newNodeShallowCopy.Annotations = nil
+	oldNodeShallowCopy.ManagedFields = nil
+	newNodeShallowCopy.ManagedFields = nil
+
+	if strings.HasPrefix(newNode.Spec.ProviderID, "gce") {
+		oldId, oldCondition := nodeutil.GetNodeCondition(&(oldNodeShallowCopy.Status), corev1.NodeNetworkUnavailable)
+		_, newCondition := nodeutil.GetNodeCondition(&(newNodeShallowCopy.Status), corev1.NodeNetworkUnavailable)
+
+		// In GCP OVN-Kubernetes modifies NodeNetworkUnavailable condition on the nodes, allow an update of the condition
+		// https://github.com/ovn-org/ovn-kubernetes/blob/e2e442133f16699671bb6564c4b8863229841fd9/go-controller/pkg/ovn/master.go#L507
+		if oldId >= 0 && newCondition != nil && !reflect.DeepEqual(oldCondition, newCondition) {
```

Comment:
heh, cool. How about looking specifically for `condition.Reason == "RouteCreated"`  as well ?

Reply:
I would like to keep that verification as is to allow for any future changes in ovn-k, we already "own" that condition so I don't see value in limiting it further.